channels/rdpgfx/server/rdpgfx_main.c

```diff
@@ -0,0 +1,1659 @@
+ * according to codecId
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static INLINE UINT16 rdpgfx_surface_command_cmdid(RDPGFX_SURFACE_COMMAND* cmd)
+{
+	if (cmd->codecId == RDPGFX_CODECID_CAPROGRESSIVE ||
+			cmd->codecId == RDPGFX_CODECID_CAPROGRESSIVE_V2)
+	{
+		return RDPGFX_CMDID_WIRETOSURFACE_2;
+	}
+
+	return RDPGFX_CMDID_WIRETOSURFACE_1;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_write_h264_metablock(wStream* s, RDPGFX_H264_METABLOCK* meta)
+{
+	UINT32 index;
+	RECTANGLE_16* regionRect;
+	RDPGFX_H264_QUANT_QUALITY* quantQualityVal;
+	UINT error = CHANNEL_RC_OK;
+	Stream_Write_UINT32(s, meta->numRegionRects); /* numRegionRects (4 bytes) */
+
+	for (index = 0; index < meta->numRegionRects; index++)
+	{
+		regionRect = &(meta->regionRects[index]);
+
+		if ((error = rdpgfx_write_rect16(s, regionRect)))
+		{
+			WLog_ERR(TAG, "rdpgfx_write_rect16 failed with error %lu!", error);
+			return error;
+		}
+	}
+
+	for (index = 0; index < meta->numRegionRects; index++)
+	{
+		quantQualityVal = &(meta->quantQualityVals[index]);
+		Stream_Write_UINT8(s, quantQualityVal->qp
+						   | (quantQualityVal->r << 6)
+						   | (quantQualityVal->p << 7)); /* qpVal (1 byte) */
+		/* qualityVal (1 byte) */
+		Stream_Write_UINT8(s, quantQualityVal->qualityVal);
+	}
+
+	return error;
+}
+
+/**
+ * Function description
+ * Write RFX_AVC420_BITMAP_STREAM structure to stream
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static INLINE UINT rdpgfx_write_h264_avc420(wStream* s,
+		RDPGFX_AVC420_BITMAP_STREAM* havc420)
+{
+	UINT error = CHANNEL_RC_OK;
+
+	if ((error = rdpgfx_write_h264_metablock(s, &(havc420->meta))))
+	{
+		WLog_ERR(TAG, "rdpgfx_write_h264_metablock failed with error %lu!",
+				 error);
+		return error;
+	}
+
+	Stream_Write(s, havc420->data, havc420->length);
+	return error;
+}
+
+/**
+ * Function description
+ * Write RDPGFX_CMDID_WIRETOSURFACE_1 or RDPGFX_CMDID_WIRETOSURFACE_2
+ * to the stream according to RDPGFX_SURFACE_COMMAND message
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_write_surface_command(wStream* s, RDPGFX_SURFACE_COMMAND* cmd)
+{
+	UINT error = CHANNEL_RC_OK;
+	RDPGFX_AVC420_BITMAP_STREAM* havc420 = NULL;
+	RDPGFX_AVC444_BITMAP_STREAM* havc444 = NULL;
+	UINT32 bitmapDataStart = 0;
+	UINT32 bitmapDataLength = 0;
+
+	if (cmd->codecId == RDPGFX_CODECID_CAPROGRESSIVE ||
+			cmd->codecId == RDPGFX_CODECID_CAPROGRESSIVE_V2)
+	{
+		/* Write RDPGFX_CMDID_WIRETOSURFACE_2 format for CAPROGRESSIVE */
+		Stream_Write_UINT16(s, cmd->surfaceId); /* surfaceId (2 bytes) */
+		Stream_Write_UINT16(s, cmd->codecId); /* codecId (2 bytes) */
+		Stream_Write_UINT32(s, cmd->contextId); /* codecContextId (4 bytes) */
+		Stream_Write_UINT8(s, cmd->format); /* pixelFormat (1 byte) */
+		Stream_Write_UINT32(s, cmd->length); /* bitmapDataLength (4 bytes) */
+		Stream_Write(s, cmd->data, cmd->length);
+	}
```

Comment:
I prefer not, because the logic model is not
```
if check1:
   return
if check2:
  return

do something
return
```
The logic model here is actually:
```
if case1:
  handle1
elseif case2:
  handle2
elseif case3:
  handle3
else
  handle4
```
Just currently we happen to have only 2 different cases. :)